### DISCUSSION TITLE
How can I speed up API responses?

### DISCUSSION THREAD

**Discussion by powerscooter** [netbox-community/netbox#14937] [Q&A]
Hey everyone,

I have around 250 devices in my Netbox Instance and therefore around 2200 interfaces.
When I call up the interfaces via API, this takes about 30 seconds. Which is far too long in practice.

Is there any way to speed up the Netbox API?

I have Netbox running on an Ubuntu server with python3.10. 

Thanks :-)

**Comment by candlerb** [comment]
Please be more specific: exactly which API endpoint are you hitting?

**Reply by powerscooter** [reply]
In this case, I am talking specifically about the API endpoint: api/dcim/interfaces

**Comment by abhi1693** [comment]
I tried loading 1000 enteries per page on the demo server and it took ~11 seconds. I dont think it's a problem with the API. Unless you can share more details about your setup and how you are making the call, it's difficult to guess what the underlying issue could be.

![image](https://github.com/netbox-community/netbox/assets/5083532/b050d169-517a-489f-8bc3-7ff19f03387a)

**Reply by powerscooter** [reply]
This matches the times I have when I call up all the interfaces in my setup.
Retrieving 2200 interfaces takes over 20 seconds and often almost 30 seconds.
So when i loading 1000 interfaces i will  have also have a time from round about 10-12 sec. 

We retrieve the interfaces using a Python script with the "Requests" library. So no special script at all.

It seems to me that the Netbox API is simply very slow.
If we now make a call of the devices/power ports and console ports in a Python script, we quickly have a call time of over 30 seconds.

We have an HP server in our lab, here Netbox has 16 CPU cores and 16 GB RAM.
Everything is installed on an Ubuntu server. Netbox is installed with python3.10
When I look at the load during a call, the server is hardly busy.
Could Postgres be the problem here?

**Reply by abhi1693** [reply]
How many objects are you fetching per page?

**Reply by powerscooter** [reply]
All interfaces - no limit

**Reply by candlerb** [reply]
The problem is very likely that the interfaces query is making additional database queries  for related objects (e.g. devices, sites etc). For example if you fetch 1000 interfaces I'm guessing it's doing 1001 database queries.

This is one of the major problems with using ORMs. Touching an attribute on an object, when the object is only partially loaded, silently causes an additional database query for that object.

To make this faster would require some analysis (tracing what additional queries are made) and optimisation (maybe `select_related` to join additional tables on the initial query)
